[BugFix] Some_problem_when_backup_restore_using_local_path (21735)

srlch opened the pull request:
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：
<!--
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #21735

## Problem Summary(Required) ：
Problem 1:
When FE uploads the backup file, FE will check the authority information. But if we use the local path, authority is undefined and FE will get an NPE problem.

Problem 2:
When we use local path for restore, BE will parse the wrong path for the remote file and let the restore fail.

Solution:
1. check authority information.
2. fix the parse problem.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [ ] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

chaoyli commented on fe/fe-core/src/main/java/com/starrocks/fs/hdfs/HdfsFsManager.java:
```diff
@@ -1238,7 +1238,10 @@ public void deletePath(String path, Map<String, String> loadProperties) throws U
+                (srcPathUri.getAuthority() != null && destPathUri.getAuthority() == null) ||
+                    (srcPathUri.getAuthority() != null && destPathUri.getAuthority() != null && 
+                        (!srcPathUri.getAuthority().trim().equals(destPathUri.getAuthority().trim())))) {
             throw new UserException(
```
The exception is not clear after change the comparison in the if branch.

srlch replied:
update the exception info

srlch commented on be/src/fs/hdfs/fs_hdfs.cpp:
```diff
@@ -371,6 +371,12 @@ Status HdfsFileSystem::iterate_dir2(const std::string& dir, const std::function<
         }
+
+        const std::string local_fs("file:/");
+        if (dir.compare(0, local_fs.length(), local_fs) == 0) {
```
dir = file://path and fileinfo[i].mName = file:path/filename，dir_size is used for extracting the file name from fileinfo[i].mName. It is the begin position from fileinfo[i].mName if we want to extract the file name.

srlch replied:
refactor the code here to make more clear

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 5 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/fs/hdfs/HdfsFsManager.java|0|5|00.00%|[1241, 1242, 1243, 1246, 1247]|

chaoyli commented on fe/fe-core/src/main/java/com/starrocks/fs/hdfs/HdfsFsManager.java:
```diff
@@ -1238,7 +1238,13 @@ public void deletePath(String path, Map<String, String> loadProperties) throws U
         WildcardURI srcPathUri = new WildcardURI(srcPath);
         WildcardURI destPathUri = new WildcardURI(destPath);
-        if (!srcPathUri.getAuthority().trim().equals(destPathUri.getAuthority().trim())) {
+        if ((srcPathUri.getAuthority() == null && destPathUri.getAuthority() != null) ||
```
```
bool srcNull = (srcPathUri.getAuthority() == null);
bool destNull = (destPathUri.getAuthority() == null);
if (srcNull ^ destNull) {

} else if (!srcNull && !destNull) {

}
```

srlch replied:
done

srlch committed:
[BugFix] Some_problem_when_backup_restore_using_local_path (21735)

Problem 1:
When fe upload the backup file, fe will check the authority information.
But if we use local path, authority is undefined and fe will get NPE problem.

Problem 2:
When we use local path for restore, BE will parse the wrong path for
the remote file and let the restore fail.

Solution:
1. check authority information.
2. fix the parse problem.

Signed-off-by: srlch <linzichao@starrocks.com>

wanpengfei-git commented:
#### [FE PR Coverage Check]

:disappointed: **fail** : 0 / 7 (00.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/fs/hdfs/HdfsFsManager.java|0|7|00.00%|[1242, 1243, 1244, 1245, 1246, 1247, 1250]|
